lib/workers/evm_server.rb

```diff
@@ -24,33 +25,80 @@ def start
     EvmDatabase.seed_primordial
     check_migrations_up_to_date
 
-    @server = MiqServer.my_server(true)
-
-    start_server
-    loop do
-      monitor
-      sleep ::Settings.server.monitor_poll.to_i_with_method
-    end
+    start_servers
+    monitor_servers
   rescue Interrupt => e
     _log.info("Received #{e.message} signal, killing server")
-    MiqServer.kill
+    kill_servers
     exit 1
   rescue SignalException => e
     _log.info("Received #{e.message} signal, shutting down server")
-    @server.shutdown_and_exit
+    stop_servers
+    exit 0
   end
 
-  ##
-  # Sets the server process' name if it is possible.
-  #
-  def set_process_title
-    Process.setproctitle(SERVER_PROCESS_TITLE) if Process.respond_to?(:setproctitle)
+  def start_servers
+    refresh_servers_to_monitor
+    for_each_server { start_server }
+  end
+
+  def monitor_servers
+    loop do
+      refresh_servers_to_monitor
+      for_each_server { monitor }
+      sleep ::Settings.server.monitor_poll.to_i_with_method
+    end
+  end
+
+  def stop_servers
+    for_each_server { @server.shutdown }
+  end
+
+  def kill_servers
+    for_each_server do
+      @server.kill_all_workers
+      @server.update(:stopped_on => Time.now.utc, :status => "killed", :is_master => false)
+    end
+  end
+
+  def refresh_servers_to_monitor
+    # Add the server object to our list and start it if we're not monitoring it already
+    servers_from_db.each do |db_server|
+      unless monitoring_server?(db_server)
+        servers_to_monitor << db_server
+        impersonate_server(db_server)
+        start_server
+      end
+    end
+
+    # Remove and shutdown a server if we're monitoring it and it is no longer in the database
+    servers_to_monitor.delete_if do |monitor_server|
+      servers_from_db.none? { |db_server| db_server.id == monitor_server.id }.tap do |should_delete|
+        monitor_server.shutdown if should_delete
+      end
+    end
```

Comment:
Pretty sure this is equivalent and I think easier to follow.

```ruby
def refresh_servers_to_monitor
  servers_to_start = servers_from_db - servers_to_monitor
  servers_to_stop = servers_to_monitor - servers_from_db

  servers_to_start.each do |s|
    servers_to_monitor << s
    impersonate_server(s)
    start_server
  end

  servers_to_delete.each do |s|
    servers_to_monitor.delete(s)
    s.shutdown
  end
end
```

This relies on ActiveRecord::Base#== being based on `id` only which simplifies the code. Additionally, this ends up dropping the need for `monitoring_server?`.

Reply:
Incidentally, if servers_to_monitor becomes a `Set`, then these operations become even more performant.

Reply:
> This relies on ActiveRecord::Base#== being based on id only

This is what I wasn't sure about. I didn't know if it compared all the attributes, which would cause an issue here as those would almost definitely differ.